fixieai/agents/agent_func.py

```diff
@@ -232,26 +307,93 @@ def create(
-    elif isinstance(result, str):
-        yield api.AgentResponse(message=api.Message(result))
-    elif isinstance(result, collections.abc.Iterable):
-        for value in result:
-            yield from _adapt_func_result(value, allow_multiple)
-
-            if not allow_multiple:
-                break
-    else:
-        raise TypeError(
-            f"The func result was type {type(result)}, but must be a str, Message, AgentResponse, or iterable."
+class AgentCorpusFunc(
+    AgentFunc[
+        corpora.CorpusRequest,
+        corpora.CorpusResponse,
+        corpora.CorpusResponse,
+    ]
+):
+    """A Python function that can be invoked by Fixie to load a custom corpus.
+
+    Wrapped functions must take one `corpora.CorpusRequest` argument (or no
+    arguments) and return a `corpora.CorpusResponse`.
+    """
+
+    def adapt_result(
+        self, result: CorpusResponse
+    ) -> Generator[CorpusResponse, None, None]:
+        if isinstance(result, CorpusResponse):
+            yield result
+        else:
+            raise TypeError(
+                f"The func result was type {type(result)}, but must be a CorpusResponse."
+            )
+
+    def adapt_exception(self, e: exceptions.AgentException) -> CorpusResponse:
+        logging.warn(f"Corpus loading threw an exception: {e}", stack_info=True)
+        return corpora.CorpusResponse()
+
+    @classmethod
+    def create(
+        cls,
+        func: Callable[[corpora.CorpusRequest], corpora.CorpusResponse],
+    ) -> AgentCorpusFunc:
+        """Constructs an AgentCorpusFunc from a Python function.
+
+        Args:
+            func: the Python function to wrap
+        """
+
+        if not inspect.isfunction(func):
+            raise TypeError(
+                f"Registered function {func!r} is not a function, but a {type(func)!r}."
+            )
+        signature = inspect.signature(func)
+        func_name = func.__name__
+        params = signature.parameters
+
+        # Validate that there are no var args (*args or **kwargs).
+        if any(
+            param.kind in (param.VAR_KEYWORD, param.VAR_POSITIONAL)
+            for param in params.values()
+        ):
+            raise TypeError(
+                f"Registered function {func_name} cannot accept variable args: {params!r}."
+            )
+
+        if len(params) > 1:
+            raise TypeError(
+                f"Registered function {func_name} must have one or zero arguments."
+            )
+        param = next(iter(params.keys())) if params else None
+        bound_request_mapper = (
+            _bind_argument_mapper(
+                func,
+                {param},
+                [
+                    (corpora.CorpusRequest, lambda request, claims: request),
+                    (param, lambda request, claims: request),
+                ],
+            )
+            if param
+            else None
+        )
+
+        if bool(params) != bool(bound_request_mapper):
```

Comment:
A perhaps-simpler alternative to going through the `_bind_argument_mapper` machinery could be to directly validate the type annotation of the parameter, but consistency with the other function is fine too.

Reply:
I initially wasn't sure this was going to be simpler, but I think you're right. Done.